ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -61,180 +77,363 @@ peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeou
+    increaseEstablishedPeers
+      mode ubp targets
+      =
+      targets {
+        targetNumberOfEstablishedPeers =
+          case (mode, ubp) of
+            (ChurnModeBulkSync, UseBootstrapPeers _) ->
+              min (targetNumberOfActivePeers targets + 1)
+                  (targetNumberOfEstablishedPeers base)
+            _  -> targetNumberOfEstablishedPeers base
+      }
+
+    checkEstablishedPeersIncreased :: CheckPeerSelectionCounters
+    checkEstablishedPeersIncreased
+      PeerSelectionCounters { numberOfEstablishedPeers }
+      PeerSelectionTargets { targetNumberOfEstablishedPeers }
+      =
+      numberOfEstablishedPeers >= targetNumberOfEstablishedPeers
+
+
+    increaseEstablishedBigLedgerPeers
+      :: ModifyPeerSelectionTargets
+    increaseEstablishedBigLedgerPeers
+      targets
+      =
+      targets { targetNumberOfEstablishedBigLedgerPeers = targetNumberOfEstablishedBigLedgerPeers base }
+
+    checkEstablishedBigLedgerPeersIncreased
+      :: CheckPeerSelectionCounters
+    checkEstablishedBigLedgerPeersIncreased
+      PeerSelectionCounters { numberOfEstablishedBigLedgerPeers }
+      PeerSelectionTargets { targetNumberOfEstablishedBigLedgerPeers }
+      =
+      numberOfEstablishedBigLedgerPeers >= targetNumberOfEstablishedBigLedgerPeers
+
+
+    decreaseEstablishedPeers
+      :: ChurnMode -> UseBootstrapPeers
+      -> ModifyPeerSelectionTargets
+    decreaseEstablishedPeers mode ubp targets =
+      targets {
+        targetNumberOfEstablishedPeers =
+          case (mode, ubp) of
+             (ChurnModeBulkSync, UseBootstrapPeers _) ->
+                 min (targetNumberOfActivePeers targets)
+                     (targetNumberOfEstablishedPeers base - 1)
+             _ -> decrease (targetNumberOfEstablishedPeers base - targetNumberOfActivePeers base)
+                    + targetNumberOfActivePeers base
+      }
+
+    checkEstablishedPeersDecreased
+      :: CheckPeerSelectionCounters
+    checkEstablishedPeersDecreased
+      PeerSelectionCounters { numberOfEstablishedPeers }
+      PeerSelectionTargets { targetNumberOfEstablishedPeers }
+      =
+      numberOfEstablishedPeers <= targetNumberOfEstablishedPeers
+
+
+    increaseActiveBigLedgerPeers
+      :: ChurnMode -> ModifyPeerSelectionTargets
+    increaseActiveBigLedgerPeers mode targets =
+      targets {
+        -- TODO: when chain-skipping will be implemented and chain-sync client
+        -- will take into account big ledger peers, we don't need pattern
+        -- match on the churn mode, but use
+        -- `targetNumberOfActiveBigLedgerPeers` (issue #4609).
+        targetNumberOfActiveBigLedgerPeers =
+          case mode of
+            ChurnModeNormal ->
+              targetNumberOfActiveBigLedgerPeers base
+            ChurnModeBulkSync ->
+              min 1 (targetNumberOfActiveBigLedgerPeers base)
+      }
+
+    checkActiveBigLedgerPeersIncreased
+      :: CheckPeerSelectionCounters
+    checkActiveBigLedgerPeersIncreased
```

Comment:
see `checkActivePeersDecreased` comment.

Reply:
Done.

Reply:
Yeah, that was intended for `checkActiveBigLedgerPeersDecreased`